File: mvcc_wal.cpp
#include <vector>
#include <iostream>
#include <string>
#include <unistd.h>
#include <fcntl.h>

#ifdef __APPLE__
#include <sys/fcntl.h>
#endif

// MVCC
struct Version {
    int value;
    int beginTx;
    int endTx; // 0 = infinity
};

// WAL
class WAL {
    int fd;

public:
    WAL(const char* path) {
        fd = open(
            path,
            O_CREAT | O_APPEND | O_WRONLY,
            0644
        );
        if (fd < 0) {
            perror("open");
            _exit(1);
        }
    }

    void force_flush() {
#ifdef __APPLE__
        // Strongest possible flush on macOS
        if (fcntl(fd, F_FULLFSYNC) == -1) {
            perror("F_FULLFSYNC");
            _exit(1);
        }
#else
        // Linux / others
        if (fsync(fd) == -1) {
            perror("fsync");
            _exit(1);
        }
#endif
    }

    void log(const std::string& record) {
        write(fd, record.c_str(), record.size());
        write(fd, "\n", 1);
        force_flush();  // FORCE DURABILITY
    }
};

// Store
class Store {
    std::vector<Version> versions;
    int nextTx = 1;
    WAL wal;

public:
    Store() : wal("wal_cpp.log") {
        versions.reserve(1024);
    }

    int begin() {
        int tx = nextTx++;
        wal.log("BEGIN " + std::to_string(tx));
        return tx;
    }

    int read(int tx) {
        for (int i = versions.size() - 1; i >= 0; i--) {
            auto &v = versions[i];
            if (v.beginTx <= tx && (v.endTx == 0 || v.endTx > tx)) {
                return v.value;
            }
        }
        return 0;
    }

    void write(int tx, int value) {
        wal.log("WRITE " + std::to_string(tx) + " " + std::to_string(value));

        for (int i = versions.size() - 1; i >= 0; i--) {
            auto &v = versions[i];
            if (v.beginTx <= tx && (v.endTx == 0 || v.endTx > tx)) {
                v.endTx = tx;
                break;
            }
        }

        versions.push_back({value, tx, 0});
    }
};


int main() {
    Store store;
    const int N = 1000;

    for (int i = 0; i < N; i++) {
        int tx = store.begin();
        store.read(tx);
        store.write(tx, i);
    }

    std::cout << "C++ MVCC & WAL done: "
              << store.read(store.read(N)) << "\n";
}

File: mvcc_wal.rs
use std::fs::{OpenOptions, File};
use std::io::Write;

// MVCC
struct Version {
    value: i32,
    begin_tx: i32,
    end_tx: i32, // 0 = infinity
}

// WAL
struct WAL {
    file: File,
}

impl WAL {
    fn new(path: &str) -> Self {
        let file = OpenOptions::new()
            .create(true)
            .append(true)
            .write(true)
            .open(path)
            .unwrap();
        WAL { file }
    }

    fn log(&mut self, record: &str) {
        writeln!(self.file, "{}", record).unwrap();
        self.file.sync_all().unwrap(); // fsync equivalent
    }
}

// Store
struct Store {
    versions: Vec<Version>,
    next_tx: i32,
    wal: WAL,
}

impl Store {
    fn new() -> Self {
        Store {
            versions: Vec::with_capacity(1024),
            next_tx: 1,
            wal: WAL::new("wal_rust.log"),
        }
    }

    fn begin(&mut self) -> i32 {
        let tx = self.next_tx;
        self.next_tx += 1;

        // WAL: BEGIN
        self.wal.log(&format!("BEGIN {}", tx));

        tx
    }

    fn read(&self, tx: i32) -> i32 {
        for v in self.versions.iter().rev() {
            if v.begin_tx <= tx && (v.end_tx == 0 || v.end_tx > tx) {
                return v.value;
            }
        }
        0
    }

    fn write(&mut self, tx: i32, value: i32) {
        // WAL: WRITE
        self.wal.log(&format!("WRITE {} {}", tx, value));

        // Apply MVCC update
        for v in self.versions.iter_mut().rev() {
            if v.begin_tx <= tx && (v.end_tx == 0 || v.end_tx > tx) {
                v.end_tx = tx;
                break;
            }
        }

        self.versions.push(Version {
            value,
            begin_tx: tx,
            end_tx: 0,
        });
    }
}

fn main() {
    let mut store = Store::new();
    const N: i32 = 1_000;

    for i in 0..N {
        let tx = store.begin();
        store.read(tx);
        store.write(tx, i);
    }

    println!("Rust MVCC & WAL done: {}", store.read(store.next_tx - 1));
}
